### PR TITLE
Treat ICU OOMs as Chakra OOMs to help crawler better understand how to bucketize

### PR DESCRIPTION
**Pull request by jackhorton**

### CODE REVIEW

**Comment by jefgen** [comment]
This might be okay as a temporary mitigation, but I would like to better understand the failures that are occurring.

**Comment by jackhorton** [comment]
Sure. For what its worth, I also realized that I can probably check for this only in GetLocaleData and IsXLocaleAvailable, since those are the first functions that get called in the normal creation flow for Intl objects, and so if ICU is going to fail to load the data file, that's the only time it should happen.

**Comment by jefgen** [comment]
Just a thought, but can you perhaps call `u_init()` first, before calling any ICU functions?
This function should attempt to load the memory-mapped data file into memory, and you can check the status for errors to see if that is really the source of the issues.
API ref in case you are curious:
http://icu-project.org/apiref/icu4c/uclean_8h.html#ab6a4215b45b2162889aa1ff01e491d4c

**Comment by jackhorton** [comment]
"u_init() will attempt to load some part of ICU's data … A successful invocation of u_init() does not, however, guarantee that all ICU data is accessible."

Any clarification on that? Why would it not be a guarantee?

**Comment by jefgen** [comment]
With ICU you can totally have some of the data broken out into separate ".res" files that might sit along side the "main" .dat file. This technique is used for some of the timezone related files, in order to more easily update them and service them. 
Additionally, somebody might have decided to split up other parts of ICU into .res files for other components like code pages, etc.
There is no way that the API can guarantee that *all* of the data files will be loaded, it can only say that  some of them would be. (Hopefully this makes sense?)

**Comment by jackhorton** [comment]
@jefgen Updated to use u_init on initialization of Intl

**Comment by jefgen** [comment]
Thanks @jackhorton this LGTM.

Just to add a few notes, for future reference: 
From stepping through the debugger and modifying the registers so as to lie about the return value from  KERNELBASE!CreateFile2 , I am able reproduce having the status == "U_MISSING_RESOURCE_ERROR" when attempting to call ucol_getKeywordValuesForLocale().

ICU is theoretically supposed to be able to function without a data file at all, and with that in mind the error code U_MISSING_RESOURCE_ERROR isn’t exactly "wrong" per-se, though it is definitely not how I would think of things.

From stepping through the debugger and modifying the registers, I can confirm that if you do call the  u_init()  function and the data file fails to be mapped into memory then you will indeed get back U_FILE_ACCESS_ERROR  -- which could be directly translated to OOM in the Chakra code.

In short, I think that perhaps some sort of ‘lazy-init’ call to the  u_init()  function before you make any ICU API calls might be the best approach for this issue.

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp [hunk 1]
```diff
@@ -454,6 +454,24 @@ namespace Js
...
+            // u_init will cause the data file to be loaded, and if we don't have enough memory to do so, we can throw OutOfMemory here.
+            // This is to protect against spurious U_MISSING_RESOURCE_ERRORs and U_FILE_ACCESS_ERRORs coming from early-lifecycle
+            // functions that require ICU data.
+            // See OS#16897150, OS#16896933, and others relating to bad statuses returned by GetLocaleData and IsLocaleAvailable
+            UErrorCode status = U_ZERO_ERROR;
+            u_init(&status);
+            if (status == U_MEMORY_ALLOCATION_ERROR || status == U_FILE_ACCESS_ERROR || status == U_MISSING_RESOURCE_ERROR)
+            {
+                // Trace that this happens in case there are build system changes that actually cause the data file to be not found
+                INTL_TRACE("Could not initialize ICU - u_init returned status %S", u_errorName(status));
```
Comment (jefgen) [inline lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: Thanks for adding a trace here as well. It might be helpful in the future to know if the error code is File Access error or memory allocation error.

**Comment by jackhorton** [comment]
@dotnet-bot test OSX static_osx_osx_release please
test OSX static_osx_osx_test please
